specs/core/0_beacon-chain.md

```diff
@@ -1726,47 +1726,41 @@ Run the following function:
 
     # Process finalizations
-    bitfield = state.justification_bitfield
-    current_epoch = get_current_epoch(state)
-    # The 2nd/3rd/4th most recent epochs are all justified, the 2nd using the 4th as source
-    if (bitfield >> 1) % 8 == 0b111 and state.previous_justified_epoch == current_epoch - 3:
-        new_finalized_epoch = state.previous_justified_epoch
-    # The 2nd/3rd most recent epochs are both justified, the 2nd using the 3rd as source
-    if (bitfield >> 1) % 4 == 0b11 and state.previous_justified_epoch == current_epoch - 2:
-        new_finalized_epoch = state.previous_justified_epoch
-    # The 1st/2nd/3rd most recent epochs are all justified, the 1st using the 3rd as source
-    if (bitfield >> 0) % 8 == 0b111 and state.current_justified_epoch == current_epoch - 2:
-        new_finalized_epoch = state.current_justified_epoch
-    # The 1st/2nd most recent epochs are both justified, the 1st using the 2nd as source
-    if (bitfield >> 0) % 4 == 0b11 and state.current_justified_epoch == current_epoch - 1:
-        new_finalized_epoch = state.current_justified_epoch
-
-    # Update state jusification/finality fields
-    state.previous_justified_epoch = state.current_justified_epoch
-    state.previous_justified_root = state.current_justified_root
-    if new_justified_epoch != state.current_justified_epoch:
-        state.current_justified_epoch = new_justified_epoch
-        state.current_justified_root = get_block_root(state, get_epoch_start_slot(new_justified_epoch))
-    if new_finalized_epoch != state.finalized_epoch:
-        state.finalized_epoch = new_finalized_epoch
-        state.finalized_root = get_block_root(state, get_epoch_start_slot(new_finalized_epoch))
+    # The 2nd/3rd/4th most recent epochs are justified, the 2nd using the 4th as source
+    if (bitfield >> 1) % 8 == 0b111 and previous_justified_epoch == get_current_epoch(state) - 3:
+        state.finalized_epoch = previous_justified_epoch
+    # The 2nd/3rd most recent epochs are justified, the 2nd using the 3rd as source
+    if (bitfield >> 1) % 4 == 0b11 and previous_justified_epoch == get_current_epoch(state) - 2:
+        state.finalized_epoch = previous_justified_epoch
+    # The 1st/2nd/3rd most recent epochs are justified, the 1st using the 3rd as source
+    if (bitfield >> 0) % 8 == 0b111 and current_justified_epoch == get_current_epoch(state) - 2:
+        state.finalized_epoch = current_justified_epoch
+    # The 1st/2nd most recent epochs are justified, the 1st using the 2nd as source
+    if (bitfield >> 0) % 4 == 0b11 and current_justified_epoch == get_current_epoch(state) - 1:
+        state.finalized_epoch = current_justified_epoch
+    if state.finalized_epoch in (state.previous_justified_epoch, state.current_justified_epoch):
+        state.finalized_root = get_block_root(state, get_epoch_start_slot(state.finalized_epoch))
```

Comment:
The test fails in the `test_no_exit_churn_too_long_since_change` case, where we test moving state forward `PERSISTENT_COMMITTEE_PERIOD` epochs.
https://github.com/ethereum/eth2.0-specs/blob/7840d29f2d1f986fea0a32184c4eaad177de145f/tests/phase0/test_sanity.py#L285-L286

with 
```
state = <build.phase0.utils.minimal_ssz.SSZType.<locals>.SSZObject object at 0x119b07438>, slot = 4294967296

    def get_block_root(state: BeaconState,
                       slot: Slot) -> Bytes32:
        """
        Return the block root at a recent ``slot``.
        """
>       assert slot < state.slot <= slot + SLOTS_PER_HISTORICAL_ROOT
E       AssertionError
```

Because the `if state.finalized_epoch in (state.previous_justified_epoch, state.current_justified_epoch)` condition is loose than `new_finalized_epoch != state.finalized_epoch`, it's possible that the `finalized_epoch` doesn't change but this line is executed: the given `finalized_slot` is from long time ago, so it can't be access from `state`. (i.e., `state.slot > finalized_slot + SLOTS_PER_HISTORICAL_ROOT`).


I suppose the easy solution is adding `new_finalized_epoch` back.

Reply:
Thank you :)